crates/wasmtime/src/component/func/host.rs

```diff
@@ -224,7 +224,10 @@ where
     lift.enter_call();
     let params = storage.lift_params(&mut lift, param_tys)?;
 
+    cx.0.call_hook(CallHook::CallingHost)?;
     let ret = closure(cx.as_context_mut(), params)?;
+    cx.0.call_hook(CallHook::ReturningFromHost)?;
```

Comment:
The error semantics here are a bit subtle I think -- to match core wasm the `ReturningFromHost` hook should be called no matter what regardelss of whether `closure` returned an error or not.

Additionally the `ReturningFromHost` hook should be delayed to the end of the function because lowering may invoke realloc in the guest which is another instance of entering wasm and I think `ReturningFromHost` followed by `CallWasm` might be a bit unexpected.